{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# IDENTIFICATION OF NOVEL CLASSES OF NEOANTIGENS IN CANCER | Tumor specific genes selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Data preparation\n",
    "\n",
    "This first cell should be modified according to the data that is going to be used. It is only available for datasets with paired samples per patient: normal and tumor. \n",
    "\n",
    "The **PROJECT** variable should be changed according to the GEO identifier.\n",
    "\n",
    "From the GEO website, the *SRR_Acc_List.txt* and *SraRunTable.txt* files should be manually downloaded and save in a directory. This directory should be specified in **SRR** variable.\n",
    "\n",
    "The pipeline is developed with the intention of running the most computationally expensive programs in a cluster. \n",
    "In this case, a Gluster File System has been used. The code to run on a cluster may need to be adapted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import os,re,shutil,glob,openpyxl\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from gtfparse import read_gtf\n",
    "from matplotlib_venn import venn2, venn2_circles, venn2_unweighted\n",
    "from matplotlib import pyplot as plt\n",
    "from IPython.display import Image\n",
    "\n",
    "PROJECT=\"GSE193567\"\n",
    "\n",
    "DIR=os.path.join(\"data\",PROJECT)\n",
    "\n",
    "try:\n",
    "    os.makedirs(DIR) #path where to store all the itermediate steps and outputs of the pipeline\n",
    "except:\n",
    "    print(\"Directory for %s already exists\" %PROJECT)\n",
    "    \n",
    "CLUSTERDIR=\"/users/genomics/marta\" #path where to run and store things that run in a cluster\n",
    "SRR=\"/projects_eg/datasets/\"+PROJECT # path where SRR_Acc_List.txt and SraRunTable.txt are stored. It should be inside a folder named with GEO accession\n",
    "SRR_ACC=os.path.join(SRR,\"SRR_Acc_List.txt\") \n",
    "SRA=os.path.join(SRR,\"SraRunTable.txt\")\n",
    "\n",
    "FASTQDIR=os.path.join(DIR,\"fastq_files\") #path where to store fastq files\n",
    "try:\n",
    "    os.mkdir(FASTQDIR)\n",
    "except:\n",
    "    print(\"Fastq_files directory exists\")\n",
    "    \n",
    "shutil.copy(SRR_ACC, os.path.join(FASTQDIR,\"SRR_Acc_List.txt\"))\n",
    "shutil.copy(SRA, os.path.join(FASTQDIR,\"SraRunTable.txt\"))\n",
    "\n",
    "GENOMEDIR=\"genomes\"\n",
    "\n",
    "try:\n",
    "    os.makedirs(os.path.join(DIR,\"analysis\"))\n",
    "    os.makedirs(os.path.join(DIR,\"results\"))\n",
    "    #os.makedirs(os.path.join(DIR,\"scripts\"))\n",
    "except:\n",
    "    print(\"Directory exists\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "require(tidyr)\n",
    "require(dplyr)\n",
    "require(rtracklayer)\n",
    "#library(purrr)\n",
    "require(ggplot2)\n",
    "require(RColorBrewer)\n",
    "require(devtools)\n",
    "require(stringr)\n",
    "require(edgeR)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get a three column file with patient_id normal_id tumor_id for latter usage "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(os.path.join(FASTQDIR.split(\"/fastq_files\")[0],\"SraRunTable.txt\"))\n",
    "metadata = metadata[['Run','Individual','tissue']]\n",
    "\n",
    "normal = metadata[metadata['tissue'] == \"non-tumor\"]\n",
    "normal = normal[['Individual','Run']]\n",
    "\n",
    "tumor = metadata[metadata['tissue'] == \"tumor\"]\n",
    "tumor = tumor[['Individual','Run']].rename(columns ={'Run' : 'Run_t'})\n",
    "\n",
    "patients = pd.merge(normal, tumor, on=['Individual'])\n",
    "patients['Individual'] = patients['Individual'].str.split(' ').str[1]\n",
    "patients.to_csv(os.path.join(DIR,\"results/patients.csv\"),index=False, header=False)\n",
    "patients_summary = os.path.join(DIR,\"results/patients.csv\")\n",
    "\n",
    "patients_id=list(patients.iloc[:,0])\n",
    "normal_id=list(patients.iloc[:,1])\n",
    "tumor_id=list(patients.iloc[:,2])\n",
    "\n",
    "patients"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 07.Quantification with featureCounts\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DIR\"\n",
    "\n",
    "mkdir $1/analysis/07_quantification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$PROJECT\" \"$CLUSTERDIR\" \"$DIR\" \"$patients_summary\"\n",
    "\n",
    "######################################DONE IN CLUSTER###############################################\n",
    "\n",
    "sbatch $3/scripts/2_tumor_specific/loop_featureCounts.sh $1 $2 $3 $4 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Process featureCounts output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DIR\"\n",
    "\n",
    "mkdir $1/results/plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i patients_summary,DIR,patients\n",
    "\n",
    "############PATIENTS INFO#############\n",
    "colnames(patients) <- c(\"patient\", \"normal\", \"tumor\")\n",
    "patients_full <- patients %>% pivot_longer(cols = !patient, names_to = \"normal_tumor\", values_to = \"sample\")\n",
    "\n",
    "###To wrangle table of counts\n",
    "for (i in 1:nrow(patients)) {\n",
    "    #read and process table of counts\n",
    "    fc_res <- read.table(paste0(DIR,\"/analysis/07_quantification/\",patients[i,1],\"/\",patients[i,1],\"_featureCounts.txt\"), header=T, fill=T) \n",
    "\n",
    "    table_of_counts <- fc_res[, -c(1:6)]\n",
    "    row.names(table_of_counts) <- fc_res$Geneid\n",
    "    sample_names <- str_sub(colnames(table_of_counts),-40,-30)\n",
    "    colnames(table_of_counts) <- sample_names\n",
    "    table_of_counts$gene_id <- fc_res$Geneid\n",
    "    table_of_counts <- table_of_counts[, c(3, 1, 2)]\n",
    "    names(table_of_counts)[1] <- \"gene_id\"\n",
    "    write.csv(table_of_counts, file=paste0(DIR,\"/analysis/07_quantification/\",patients[i,1],\"/\",patients[i,1],\"_table_of_counts.csv\"), row.names = FALSE, quote = FALSE)\n",
    "\n",
    "\n",
    "    featureLength <- dplyr::select(fc_res, Length, Geneid)\n",
    "\n",
    "    ###FROM COUNTS TO FPKM\n",
    "    y <- DGEList(counts=table_of_counts[,c(2,3)],genes=data.frame(Length=featureLength$Length))\n",
    "    y <- calcNormFactors(y)\n",
    "    RPKM_raw <- as.data.frame(rpkm(y)) \n",
    "    RPKM <- RPKM_raw\n",
    "    RPKM$gene_id <- rownames(RPKM)\n",
    "    RPKM <- RPKM[, c(3, 1, 2)]\n",
    "    novel_RPKM <- RPKM[grepl(\"STRG\", RPKM$gene_id),] \n",
    "    known_RPKM <- RPKM[!grepl(\"STRG\", RPKM$gene_id),] \n",
    "\n",
    "    write.csv(RPKM, file=paste0(DIR,\"/analysis/07_quantification/\",patients[i,1],\"/\",patients[i,1],\"_table_of_counts_RPKM.csv\"), row.names = FALSE, quote = FALSE)\n",
    "\n",
    "    large_RPKM <- RPKM %>% pivot_longer(cols = !gene_id, names_to = \"sample\", values_to = \"RPKM\")\n",
    "\n",
    "    df <- merge(large_RPKM, patients_full, by=\"sample\", all=T)\n",
    "    df$logRPKM <- log10(df$RPKM)\n",
    "    df <- df %>% mutate(known_novel = case_when(grepl(\"STRG\", gene_id) ~ \"novel\",\n",
    "                                                        !grepl(\"STRG\", gene_id) ~ \"known\"))\n",
    "    colnames(featureLength) <- c(\"Length\", \"gene_id\")\n",
    "    df <- merge(df, featureLength, by=\"gene_id\")\n",
    "\n",
    "    ###length\n",
    "    RPKM_length <- merge(RPKM, featureLength, by=\"gene_id\")\n",
    "    novel_genes <- df[df$known_novel == \"novel\",]\n",
    "    known_genes <- df[df$known_novel == \"known\",]\n",
    "\n",
    "    novel_genes_RPKM <- RPKM_length[grepl(\"STRG\", RPKM_length$gene_id),] \n",
    "    known_genes_RPKM <- RPKM_length[!grepl(\"STRG\", RPKM_length$gene_id),] \n",
    "\n",
    "    ###get those with length > 300\n",
    "    novel_genes_300 <- novel_genes_RPKM[novel_genes_RPKM$Length >= 300,]\n",
    "    dim(novel_genes_300)  \n",
    "\n",
    "    filtered_RPKM <- rbind(known_genes_RPKM, novel_genes_300)\n",
    "\n",
    "    filtered_df <- df %>% filter(df$gene_id %in% filtered_RPKM$gene_id)\n",
    "    write.csv(filtered_RPKM, file=paste0(DIR,\"/analysis/07_quantification/\",patients[i,1],\"/\",patients[i,1],\"_table_of_counts_300kb_RPKM.csv\"), sep=\",\", row.names = FALSE, quote = FALSE)\n",
    "\n",
    "    \n",
    "    ###########PLOTS\n",
    "    \n",
    "    #ggplot(df, aes(x=Length, fill=known_novel)) +\n",
    "    #geom_histogram(position = 'identity') +\n",
    "    #  scale_fill_manual(values=c(\"#636363\", \"#fa9fb5\")) +\n",
    "    #scale_x_continuous(limits = c(0, 1000)) +\n",
    "    #  ggtitle(paste0(\"Histogram of all gene lengths \",patients[i,1])) +\n",
    "    #ylab(\"\") +\n",
    "    #xlab(\"Gene length\")\n",
    "    #ggsave(paste0(DIR,\"/results/plots/\",patients[i,1],\"_gene_length_histogram.png\"))\n",
    "\n",
    "    #ggplot(df, aes(x=Length, fill=known_novel)) +\n",
    "    #geom_histogram(position = 'identity') +\n",
    "    #  scale_fill_manual(values=c(\"#636363\", \"#fa9fb5\")) +\n",
    "    #scale_x_continuous(limits = c(0, 1000)) +\n",
    "    #scale_y_continuous(limits = c(0, 3000)) +\n",
    "    #  ggtitle(paste0(\"Histogram of all gene lengths \",patients[i,1],\" (zoom)\")) +\n",
    "    #ylab(\"\") +\n",
    "    #xlab(\"Gene length\")\n",
    "    #ggsave(paste0(DIR,\"/results/plots/\",patients[i,1],\"_gene_length_histogram_zoom.png\"))\n",
    "\n",
    "    \n",
    "    #ggplot(df, aes(x=logRPKM, color=known_novel)) +\n",
    "    #    geom_density() +\n",
    "    #      scale_color_manual(\"Known or Novel genes\", values=c(\"#636363\", \"#fa9fb5\")) +\n",
    "    #    ggtitle(paste0(\"FPKM distribution \",patients[i,1])) +\n",
    "    #    theme(plot.title = element_text(size=10, face=\"bold\"))\n",
    "    #ggsave(paste0(DIR, \"/results/plots/\",patients[i,1],\"_FPKM_distribution.png\"))\n",
    "\n",
    "    #ggplot(filtered_df, aes(x=logRPKM, color=known_novel)) +\n",
    "    #    geom_density() +\n",
    "    #      scale_color_manual(\"Known or Novel genes\", values=c(\"#636363\", \"#fa9fb5\")) +\n",
    "    #    ggtitle(paste0(\"FPKM distribution \",patients[i,1],\" >300kb\")) +\n",
    "    #    theme(plot.title = element_text(size=10, face=\"bold\"))\n",
    "    #ggsave(paste0(DIR, \"/results/plots/\",patients[i,1],\"_FPKM_distribution_300kb.png\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make summary file after quantification and after filtering by length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DIR\" \"$patients_summary\"\n",
    "\n",
    "OUT=$1/results/quantification_num_genes_patient.txt\n",
    "if [ -f \"$OUT\" ] ; then\n",
    "    rm \"$OUT\"\n",
    "fi\n",
    "OUT2=$1/results/quantification_num_genes_patient_300kb.txt\n",
    "if [ -f \"$OUT2\" ] ; then\n",
    "    rm \"$OUT2\"\n",
    "fi\n",
    "sed 1d $2 | while IFS=, read patient normal tumor; do\n",
    "echo -e ${patient}\"\\tTRANSCRIPTS\\t\"$(wc -l < $1/analysis/07_quantification/${patient}/${patient}_table_of_counts_RPKM.csv) >> $OUT\n",
    "echo -e ${patient}\"\\tANNOTATED TRANSCRIPTS\\t\"$(grep -v 'STRG' $1/analysis/07_quantification/${patient}/${patient}_table_of_counts_RPKM.csv | wc -l) >> $OUT\n",
    "echo -e ${patient}\"\\tNOVEL TRANSCRIPTS\\t\"$(grep 'STRG' $1/analysis/07_quantification/${patient}/${patient}_table_of_counts_RPKM.csv | wc -l) >> $OUT\n",
    "\n",
    "##300kb\n",
    "echo -e ${patient}\"\\tTRANSCRIPTS\\t\"$(wc -l < $1/analysis/07_quantification/${patient}/${patient}_table_of_counts_300kb_RPKM.csv) >> $OUT2\n",
    "echo -e ${patient}\"\\tANNOTATED TRANSCRIPTS\\t\"$(grep -v 'STRG' $1/analysis/07_quantification/${patient}/${patient}_table_of_counts_300kb_RPKM.csv | wc -l) >> $OUT2\n",
    "echo -e ${patient}\"\\tNOVEL TRANSCRIPTS\\t\"$(grep 'STRG' $1/analysis/07_quantification/${patient}/${patient}_table_of_counts_300kb_RPKM.csv | wc -l)  >> $OUT2\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 08. Tumor specific\n",
    "\n",
    "To filter what we consider as tumor specific, we apply a cutoff of FPKM lower or equal to 0,1 in normal samples and higher or equal than 1 in tumor samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(os.path.join(DIR,\"analysis/08_tumor_specific\")):\n",
    "    os.mkdir(os.path.join(DIR,\"analysis/08_tumor_specific\"))\n",
    "\n",
    "for i in patients.iloc[:,0]:\n",
    "    INDIR=DIR+\"/analysis/07_quantification/\" + i\n",
    "    OUTDIR=DIR+\"/analysis/08_tumor_specific/\" + i\n",
    "    if not os.path.exists(OUTDIR):\n",
    "        os.makedirs(OUTDIR)\n",
    "        \n",
    "    for file in os.listdir(INDIR):\n",
    "        if file.endswith(\"_table_of_counts_300kb_RPKM.csv\"):\n",
    "            filepath = os.path.join(INDIR, file)\n",
    "            outfile = OUTDIR + \"/\" + file.split(\"_\")[0] + \"_genes_N01_T1_300kb.csv\"\n",
    "            with open(filepath) as inp:\n",
    "                with open(outfile, 'w') as out:\n",
    "                    lines = inp.readlines()\n",
    "                    out.write(lines[0])\n",
    "                    del lines[0]\n",
    "                    for line in lines:\n",
    "                        line = line.rstrip()\n",
    "                        line = line.split(\",\")\n",
    "                        line[1] = float(line[1])\n",
    "                        line[2] = float(line[2])\n",
    "                        if line[1] <= 0.1 and line[2] >= 1: # expression cut - off established. FPKM <= 0.1 in normal and >= 1 in tumor\n",
    "                            listToStr = ','.join(map(str, line))\n",
    "                            out.write(listToStr + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make summary file with tumor specific genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DIR\" \"$patients_summary\"\n",
    "\n",
    "OUT=$1/results/tumor_specific_genes_300kb.txt\n",
    "if [ -f \"$OUT\" ] ; then\n",
    "    rm \"$OUT\"\n",
    "fi\n",
    "sed 1d $2 | while IFS=, read patient normal tumor; do\n",
    "echo -e ${patient}\" TRANSCRIPTS: \"$(wc -l < $1/analysis/08_tumor_specific/${patient}/${patient}_genes_N01_T1_300kb.csv) >> $OUT\n",
    "echo -e \"\\tANNOTATED TRANSCRIPTS: \"$(grep -v 'STRG' $1/analysis/08_tumor_specific/${patient}/${patient}_genes_N01_T1_300kb.csv | wc -l) >> $OUT\n",
    "echo -e \"\\tNOVEL TRANSCRIPTS: \"$(grep 'STRG' $1/analysis/08_tumor_specific/${patient}/${patient}_genes_N01_T1_300kb.csv | wc -l) >> $OUT\n",
    "\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i DIR,patients\n",
    "\n",
    "# generate different csv files with novel and annotated transcripts tumor specific\n",
    "all <- data.frame(gene=character(),\n",
    "                  RPKM = numeric(),\n",
    "                  Length=numeric(),\n",
    "                  stringsAsFactors = FALSE)\n",
    "for (i in 1:nrow(patients)) {\n",
    "    input <- read.table(paste0(DIR,\"/analysis/08_tumor_specific/\",patients[i,1],\"/\",patients[i,1],\"_genes_N01_T1_300kb.csv\"), header=TRUE, sep=\",\")\n",
    "  \n",
    "    names(input)[2] <- \"normal\"\n",
    "    input$normal <- NULL\n",
    "    input <- input %>% mutate(known_novel = case_when(grepl(\"STRG\", gene_id) ~ \"novel\",\n",
    "                                                      !grepl(\"STRG\", gene_id) ~ \"known\"))\n",
    "    #write.csv(input, file=paste0(DIR, \"/analysis/08_tumor_specific/\",patients[i,1],\"/\",patients[i,1],\"_tumor_specific_genes_01_1FPKM_300kb.csv\"), quote=FALSE, row.names = FALSE, sep=\",\")\n",
    "\n",
    "    #Select the novel ones, the patient and the sample\n",
    "    novel_tumor_specific <- input[grep(\"novel\", input$known_novel),]\n",
    "    write.table(novel_tumor_specific, file=paste0(DIR, \"/analysis/08_tumor_specific/\",patients[i,1],\"/\",patients[i,1],\"_novel_tumor_specific_genes_01_1FPKM_300kb.csv\"), quote=FALSE, row.names = FALSE, sep=\",\")\n",
    "    \n",
    "    #Select the known ones, the patient and the sample\n",
    "    known_tumor_specific <- input[grep(\"known\", input$known_novel),]\n",
    "    write.table(known_tumor_specific, file=paste0(DIR, \"/analysis/08_tumor_specific/\",patients[i,1],\"/\",patients[i,1],\"_known_tumor_specific_genes_01_1FPKM_300kb.csv\"),  quote=FALSE, row.names = FALSE, sep=\",\")\n",
    "    \n",
    "    names(input)[2] <- \"RPKM\"\n",
    "    input$patient <- patients[i,1]\n",
    "    input$logRPKM <- log10(input$RPKM)\n",
    "    all <- rbind(input,all)\n",
    " \n",
    "}\n",
    "\n",
    "####################################PLOTS####################################\n",
    "####BARPLOT known/novel tumor specific genes\n",
    "ggplot(all, aes(patient, ..count.., fill=known_novel)) +\n",
    "      geom_bar(aes(fill=known_novel), position=\"dodge\") +\n",
    "      geom_text(aes(label=..count..), stat=\"count\", size = 2, position=position_dodge(0.9)) +\n",
    "      scale_fill_manual(\"Known or Novel genes\", values=c(\"#636363\", \"#fa9fb5\")) +\n",
    "      ggtitle(paste0(\"Known/Novel tumor specific genes | FPKM 1 | 300kb\")) +\n",
    "      xlab(\"\") +\n",
    "      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.title.y = element_blank(), plot.title = element_text(face=\"bold\", size = 10))\n",
    "ggsave(paste0(DIR,\"/results/plots/tumor_specific_genes_knownnovel_01_FPKM1_300kb.png\"))\n",
    "\n",
    "####BARPLOT known/novel proportion tumor specific genes\n",
    "ggplot(all, aes(patient, ..count.., fill=known_novel)) +\n",
    "      geom_bar(position=\"fill\") +\n",
    "      scale_fill_manual(\"Known or Novel genes\", values=c(\"#636363\", \"#fa9fb5\")) +\n",
    "      ggtitle(paste0(\"Known/Novel tumor specific genes | FPKM 1 | 300kb\")) +\n",
    "      xlab(\"\") +\n",
    "      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.title.y = element_blank(), plot.title = element_text(face=\"bold\", size = 10))\n",
    "ggsave(paste0(DIR,\"/results/plots/stacked_tumor_specific_genes_knownnovel_01_FPKM1_300kb.png\"))\n",
    "\n",
    "####HISTOGRAM gene length tumor speicifc genes facet by patient\n",
    "ggplot(all, aes(x=Length, fill=known_novel)) +\n",
    "      geom_histogram() +\n",
    "      scale_fill_manual(\"Known or Novel genes\", values=c(\"#636363\", \"#fa9fb5\")) +\n",
    "      scale_x_continuous(limits = c(0, 1000)) +\n",
    "      ggtitle(paste0(\"Tumor specific gene length | FPKM 1 | 300kb\")) +\n",
    "      ylab(\"\") +\n",
    "      theme(axis.title.y = element_blank(), plot.title = element_text(face=\"bold\", size = 10)) +\n",
    "  facet_wrap(~patient, scales=\"free\")\n",
    "ggsave(paste0(DIR,\"/results/plots/tumor_specific_gene_length_01_FPKM1_300kb.png\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate mini gtf tumor specific novel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for each csv file containing tumor specific genes, either novel or annotated, we generate a gtf file. \n",
    "#3 in total: a tumor specific one, another one only with novel features and another one with annotated features\n",
    "counter=0\n",
    "for p in patients_id:\n",
    "    REF=DIR+\"/analysis/06_stringtie/stringtie_reference_annotations/\"+tumor_id[counter]+\"_reference_annotation_sorted.gtf\"\n",
    "    with open(REF) as annotations:\n",
    "        annotation_lines = annotations.readlines()\n",
    "        CSV=DIR+\"/analysis/08_tumor_specific/\"+p+\"/\" + p + \"_genes_N01_T1_300kb.csv\"\n",
    "        g_id = []\n",
    "        csv_file=pd.read_csv(CSV)\n",
    "        g_id = list(csv_file.iloc[:,0])\n",
    "        OUT=DIR+\"/analysis/08_tumor_specific/\"+p+\"/\" + p + \"_tumor_specific_genes_1FPKM_300kb.gtf\"\n",
    "        OUT_known=DIR+\"/analysis/08_tumor_specific/\"+p+\"/\" + p + \"_known_tumor_specific_genes_1FPKM_300kb.gtf\"\n",
    "        OUT_novel=DIR+\"/analysis/08_tumor_specific/\"+p+\"/\" + p + \"_novel_tumor_specific_genes_1FPKM_300kb.gtf\"\n",
    "\n",
    "\n",
    "        with open(OUT, 'w') as out:\n",
    "            with open(OUT_novel, 'w') as novel:\n",
    "                with open(OUT_known, 'w') as known:\n",
    "\n",
    "                    for line in annotation_lines:\n",
    "                        if line.startswith('#'):\n",
    "                            next\n",
    "                        else:\n",
    "                            splitted_line = line.split(\";\")\n",
    "                            ENSG = splitted_line[0]\n",
    "                            ENSG = re.findall('\"([^\"]*)\"', ENSG)\n",
    "                            if ENSG[0] in g_id:\n",
    "                                out.write(line)\n",
    "                                if ENSG[0].startswith(\"ENSG\"):\n",
    "                                    known.write(line)\n",
    "                                else:\n",
    "                                    novel.write(line)\n",
    "    counter = counter + 1\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Known**\n",
    "\n",
    "Coding or non-coding?\n",
    "\n",
    "For the annotated features, we are differentially interested in the coding and non-coding ones."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%R -i DIR,patients\n",
    "\n",
    "colnames(patients) <- c(\"patient\", \"normal\", \"tumor\")\n",
    "patients_full <- patients %>% pivot_longer(cols = !patient, names_to = \"normal_tumor\", values_to = \"sample\")\n",
    "\n",
    "full <- data.frame(gene_id = character(),\n",
    "                      gene_type = factor(),\n",
    "                      RPKM=numeric(),\n",
    "                   patient=factor(),\n",
    "                      stringsAsFactors = FALSE)\n",
    "\n",
    "for (i in 1:nrow(patients)) {\n",
    "  \n",
    "  file_CDS <- paste0(DIR, \"/analysis/08_tumor_specific/\",patients[i,1],\"/\",patients[i,1],\"_known_tumor_specific_genes_1FPKM_300kb.gtf\")\n",
    "  csv <- read.csv(paste0(DIR, \"/analysis/08_tumor_specific/\",patients[i,1],\"/\",patients[i,1],\"_known_tumor_specific_genes_01_1FPKM_300kb.csv\"))\n",
    "  names(csv)[2] <- \"RPKM\"\n",
    "  names(csv)[1] <- \"gene_id\"\n",
    "  \n",
    "  gtf <- as.data.frame(rtracklayer::import(file_CDS)) # tumor specific annotated gtf file\n",
    "  NOCDS <- gtf[gtf$gene_type != \"protein_coding\",]\n",
    "  export(NOCDS, paste0(DIR, \"/analysis/08_tumor_specific/\",patients[i,1],\"/\",patients[i,1],\"_known_tumor_specific_genes_1FPKM_300kb_NOCDS.gtf\"))\n",
    "  \n",
    "  CDS <- gtf[gtf$gene_type == \"protein_coding\",]\n",
    "  export(CDS, paste0(DIR, \"/analysis/08_tumor_specific/\",patients[i,1],\"/\",patients[i,1],\"_known_tumor_specific_genes_1FPKM_300kb_CDS.gtf\"))\n",
    "  \n",
    "  ##NOCDS\n",
    "  clean_gtf <- gtf[gtf$type == \"gene\",]\n",
    "  clean_gtf <- clean_gtf %>% select(seqnames, type,gene_id, transcript_id, gene_type, transcript_type, gene_name, transcript_name)\n",
    "  \n",
    "  df <- merge(clean_gtf, csv, by=\"gene_id\")\n",
    "  df$logRPKM <- log10(df$RPKM)\n",
    "  df <- df %>% mutate(coding = case_when(grepl(\"protein_coding\", gene_type) ~ \"coding\",\n",
    "                                                    !grepl(\"protein_coding\", gene_type) ~ \"non-coding\"))\n",
    "  df$patient <- patients[i,1]\n",
    "  full <- rbind(full, df)\n",
    "  \n",
    "\n",
    "  clean_gtf_NOCDS <- NOCDS[NOCDS$type == \"gene\",]\n",
    "  clean_gtf_NOCDS <- clean_gtf_NOCDS %>% select(seqnames, type,gene_id, gene_type, gene_name)\n",
    "  \n",
    "  df_NOCDS <- merge(clean_gtf_NOCDS, csv, by=\"gene_id\")\n",
    "  df_NOCDS$logRPKM <- log10(df_NOCDS$RPKM)\n",
    "  \n",
    "  selected_NOCDS <- df_NOCDS[grep(\"lncRNA|^processed_pseudogene\", df_NOCDS$gene_type),] #we are only interested in long non-coding RNAs and processed pseudogenes\n",
    "  \n",
    "  write.csv(selected_NOCDS, file=paste0(DIR,\"/analysis/08_tumor_specific/\",patients[i,1],\"/\",patients[i,1],\"_known_tumor_specific_genes_1FPKM_300kb_NOCDS_selected.csv\"), row.names = FALSE, quote = FALSE)\n",
    "\n",
    "    \n",
    "###################################PLOTS PER PATIENT#####################################33\n",
    "#  ggplot(df, aes(x=logRPKM, color=coding)) +\n",
    "#    geom_density() +\n",
    "#    scale_color_manual(\"Coding or non-coding genes\", values=c(\"#636363\", \"#7fcdbb\")) +\n",
    "#    ggtitle(paste0(\"FPKM distribution \",patients[i,1])) +\n",
    "#    theme(plot.title = element_text(size=10, face=\"bold\"))\n",
    "#  ggsave(paste0(DIR,\"/results/plots/\",patients[i,1],\"_coding_tumor_specific_genes_logdistribution_FPKM1_300kb.png\"))\n",
    "\n",
    "\n",
    "#  ggplot(df, aes(x=coding, fill=coding)) +\n",
    "#    geom_bar() +\n",
    "#    geom_text(aes(label=..count..), stat=\"count\", size = 2, position=position_dodge(0.9)) +\n",
    "#    scale_fill_manual(\"Coding or non-coding genes\", values=c(\"#636363\", \"#7fcdbb\")) +\n",
    "#    ggtitle(paste0(\"Coding or non-coding tumor specific genes \",patients[i,1], \" | FPKM 1 | 300kb\")) +\n",
    "#    xlab(\"\") +\n",
    "#    theme(axis.title.y = element_blank(), plot.title = element_text(face=\"bold\", size = 10))\n",
    "#  ggsave(paste0(DIR,\"/results/plots/\",patients[i,1],\"_coding_tumor_specific_genes_FPKM1_300kb.png\"))\n",
    "  \n",
    "   \n",
    "#   ggplot(selected_NOCDS, aes(x=logRPKM, color=gene_type)) +\n",
    "#    geom_density() +\n",
    "#     scale_color_manual(\"Gene type\", values=c(\"#addd8e\", \"#a6bddb\")) +\n",
    "#     ggtitle(paste0(\"FPKM distribution \",patients[i,1],\" non-coding genes\")) +\n",
    "#    theme(plot.title = element_text(size=10, face=\"bold\"))\n",
    "#  ggsave(paste0(DIR,\"/results/plots/\",patients[i,1],\"_non-coding_selected_tumor_specific_genes_logdistribution_FPKM1_300kb.png\"))\n",
    "  \n",
    "  \n",
    "#  ggplot(selected_NOCDS, aes(x=gene_type, fill=gene_type)) +\n",
    "#    geom_bar() +\n",
    "#    geom_text(aes(label=..count..), stat=\"count\", size = 2, position=position_dodge(0.9)) +\n",
    "#    scale_fill_manual(\"Gene type\", values=c(\"#addd8e\", \"#a6bddb\")) +\n",
    "#    ggtitle(paste0(\"Non-coding tumor specific genes \",patients[i,1], \" | FPKM 1 | 300kb\")) +\n",
    "#    xlab(\"\") +\n",
    "#    theme(axis.title.y = element_blank(), plot.title = element_text(face=\"bold\", size = 10))\n",
    "#  ggsave(paste0(DIR,\"/results/plots/\",patients[i,1],\"_non-coding_selected_tumor_specific_genes_FPKM1_300kb.png\"))\n",
    "  \n",
    "}\n",
    "###################################PLOTS#####################################33\n",
    "\n",
    "ggplot(full, aes(x=coding, fill=coding)) +\n",
    "  geom_bar() +\n",
    "  geom_text(aes(label=..count..), stat=\"count\", size = 2, position=position_dodge(0.9)) +\n",
    "  scale_fill_manual(\"Coding or non-coding genes\", values=c(\"#636363\", \"#7fcdbb\")) +\n",
    "  ggtitle(paste0(\"Coding or non-coding tumor specific genes | FPKM 1 | 300kb\")) +\n",
    "  xlab(\"\") +\n",
    "  theme(axis.text.x = element_blank(), axis.title.y = element_blank(), plot.title = element_text(face=\"bold\", size = 10))+\n",
    "  facet_wrap(~patient, scales = \"free\")\n",
    "ggsave(paste0(DIR,\"/results/plots/coding_tumor_specific_genes_FPKM1_300kb.png\"))\n",
    "\n",
    "ggplot(full, aes(x=patient, ..count.., fill=coding)) +\n",
    "  geom_bar(position=\"fill\") +\n",
    "  scale_fill_manual(\"Coding or non-coding genes\", values=c(\"#636363\", \"#7fcdbb\")) +\n",
    "  ggtitle(paste0(\"Coding or non-coding tumor specific genes | FPKM 1 | 300kb\")) +\n",
    "  xlab(\"\") +\n",
    "  theme(axis.text.x = element_text(angle = 45, hjust=1), axis.title.y = element_blank(), plot.title = element_text(face=\"bold\", size = 10))\n",
    "ggsave(paste0(DIR,\"/results/plots/coding_tumor_specific_genes_FPKM1_300kb_stacked.png\"))\n",
    "\n",
    "ggplot(full, aes(x=logRPKM, color=coding)) +\n",
    "  geom_density() +\n",
    "  scale_color_manual(\"Coding or non-coding genes\", values=c(\"#636363\", \"#7fcdbb\")) +\n",
    "  ggtitle(\"FPKM distribution coding/non-coding genes\") +\n",
    "  theme(plot.title = element_text(size=10, face=\"bold\")) +\n",
    "  facet_wrap(~patient, scales=\"free\")\n",
    "ggsave(paste0(DIR,\"/results/plots/coding_tumor_specific_genes_logdistribution_FPKM1_300kb.png\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generate minigtf of NOCDS selected (lncRNA and processed_pseudogenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counter=0\n",
    "for p in patients_id:\n",
    "    print(p,\" is being read\")\n",
    "    REF=DIR+\"/analysis/06_stringtie/stringtie_reference_annotations/\"+tumor_id[counter]+\"_reference_annotation_sorted.gtf\"\n",
    "    with open(REF) as annotations:\n",
    "        annotation_lines = annotations.readlines()\n",
    "        CSV=DIR+\"/analysis/08_tumor_specific/\"+p+\"/\" + p + \"_known_tumor_specific_genes_1FPKM_300kb_NOCDS_selected.csv\"\n",
    "        csv_file=pd.read_csv(CSV)\n",
    "        g_id = list(csv_file.iloc[:,0])\n",
    "        OUT_known=DIR+\"/analysis/08_tumor_specific/\"+p+\"/\" + p + \"_known_tumor_specific_genes_1FPKM_300kb_NOCDS_selected.gtf\"\n",
    "\n",
    "        with open(OUT_known, 'w') as out:\n",
    "            for line in annotation_lines:\n",
    "                if line.startswith('#'):\n",
    "                    next\n",
    "                else:\n",
    "                    splitted_line = line.split(\";\")\n",
    "                    ENSG = splitted_line[0]\n",
    "                    ENSG = re.findall('\"([^\"]*)\"', ENSG)\n",
    "                    if ENSG[0] in g_id:\n",
    "                        out.write(line)\n",
    "\n",
    "    counter = counter + 1\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#more than one transcript line per gene_id \n",
    "#these are the unique ones\n",
    "OUT=DIR+\"/results/transcripts_genes_tumor_specific.txt\"\n",
    "with open(OUT,'w') as out:\n",
    "    for p in patients_id:\n",
    "        file=DIR+\"/analysis/08_tumor_specific/\"+p+\"/\"+p+\"_known_tumor_specific_genes_1FPKM_300kb_NOCDS_selected.gtf\"\n",
    "        genes_list = []\n",
    "        with open(file) as inp:\n",
    "            lines=inp.readlines()\n",
    "            for line in lines:\n",
    "                if \"\\ttranscript\\t\" in line:\n",
    "                    splitted_line = line.split(\";\")\n",
    "                    gene_id = splitted_line[0]\n",
    "                    ENSG = re.findall('\"([^\"]*)\"', gene_id)\n",
    "                    genes_list.append(ENSG[0])\n",
    "        genes_set = set(genes_list)\n",
    "        out.write(p+\"gene\"+str(len(genes_set)))\n",
    "        out.write(p+\"transcript\"+str(len(genes_list)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get fasta of noncoding selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DIR\" \"$patients_summary\" \"$GENOMEDIR\"\n",
    "\n",
    "export PATH=/genomics/users/marta/tools/gffread-0.12.7.Linux_x86_64/:$PATH\n",
    "GENOME_FASTA=$3/GRCh38/GRCh38.primary_assembly.genome.fa\n",
    "\n",
    "while IFS=, read p normal tumor; do\n",
    "    file=$1/analysis/08_tumor_specific/${p}/${p}_known_tumor_specific_genes_1FPKM_300kb_NOCDS_selected.gtf\n",
    "    \n",
    "    #get fasta\n",
    "    gffread --attrs gene_name,transcript_type,transcript_name -w ${file%%.*}.fa -g $GENOME_FASTA $file\n",
    "\n",
    "    #count fasta sequences to confirm the process\n",
    "    echo -e $p\"\\t\"$(grep '>' ${file%%.*}.fa | wc -l)\n",
    "    \n",
    "    #replace spaces by ;\n",
    "    sed -i 's/\\ /;/g' ${file%%.*}.fa\n",
    "done < $2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we will work at gene level. For genes with more than one transcript, we keep the longest one\n",
    "for p in patients_id:\n",
    "    g_dict= dict()\n",
    "    full = dict()\n",
    "    file=DIR+\"/analysis/08_tumor_specific/\"+p+\"/\"+p+\"_known_tumor_specific_genes_1FPKM_300kb_NOCDS_selected.fa\"\n",
    "    OUT=DIR+\"/analysis/08_tumor_specific/\"+p+\"/\"+p+\"_known_tumor_specific_genes_1FPKM_300kb_NOCDS_selected_gene.fa\"\n",
    "    with open(OUT, 'w') as out:\n",
    "        for seq_record in SeqIO.parse(file, 'fasta'):  # (generator)   \n",
    "            gene_name = seq_record.id.split(\";\")[1]\n",
    "            gene_name = gene_name.split(\"=\")[1]\n",
    "            if gene_name in g_dict.keys():\n",
    "                if len(seq_record.seq) > len(g_dict[gene_name]): # get the longest transcript per gene\n",
    "                    g_dict[gene_name] = seq_record.seq\n",
    "                    full[gene_name] = seq_record.id\n",
    "                else:       \n",
    "                    next\n",
    "            else:\n",
    "                g_dict[gene_name] = seq_record.seq\n",
    "                full[gene_name] = seq_record.id\n",
    "                \n",
    "        print(len(full))\n",
    "        for identif, seq in g_dict.items():\n",
    "            out.write(\">%s\\n%s\\n\" %(full[identif], seq))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**CDS**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%bash -s \"$DIR\" \"$patients_summary\" \"$GENOMEDIR\"\n",
    "\n",
    "export PATH=/genomics/users/marta/tools/gffread-0.12.7.Linux_x86_64/:$PATH\n",
    "GENOME_FASTA=$3/GRCh38/GRCh38.primary_assembly.genome.fa\n",
    "\n",
    "cat $2 | while IFS=, read p normal tumor; do\n",
    "    file=$1/analysis/08_tumor_specific/${p}/${p}_known_tumor_specific_genes_1FPKM_300kb_CDS.gtf\n",
    "    \n",
    "    #get fasta\n",
    "    gffread --attrs gene_name,transcript_type,transcript_name -x ${file%%.*}.fa -g $GENOME_FASTA $file\n",
    "    \n",
    "    #count fasta sequences to confirm the process\n",
    "    echo -e $p\"\\t\"$(grep '>' ${file%%.*}.fa | wc -l)\n",
    "    \n",
    "    #replace spaces by ;\n",
    "    sed -i 's/\\ /;/g' ${file%%.*}.fa\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we will work at gene level. For genes with more than one transcript, we keep the longest one\n",
    "for p in patients_id:\n",
    "    g_dict= dict()\n",
    "    full = dict()\n",
    "    file=DIR+\"/analysis/08_tumor_specific/\"+p+\"/\"+p+\"_known_tumor_specific_genes_1FPKM_300kb_CDS.fa\"\n",
    "    OUT=DIR+\"/analysis/08_tumor_specific/\"+p+\"/\"+p+\"_known_tumor_specific_genes_1FPKM_300kb_CDS_gene.fa\"\n",
    "    with open(OUT, 'w') as out:\n",
    "        for seq_record in SeqIO.parse(file, 'fasta'):  # (generator)   \n",
    "            gene_name = seq_record.id.split(\";\")[1]\n",
    "            gene_name = gene_name.split(\"=\")[1]\n",
    "            if gene_name in g_dict.keys():\n",
    "                if len(seq_record.seq) > len(g_dict[gene_name]): # get longest transcript per gene\n",
    "                    g_dict[gene_name] = seq_record.seq\n",
    "                    full[gene_name] = seq_record.id\n",
    "                else:\n",
    "                    next\n",
    "            else:\n",
    "                g_dict[gene_name] = seq_record.seq\n",
    "                full[gene_name] = seq_record.id\n",
    "                \n",
    "        print(len(full))\n",
    "        for identif, seq in g_dict.items():\n",
    "            out.write(\">%s\\n%s\\n\" %(full[identif], seq))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For protein-coding genes, how many of them are shared between patients ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DIR\"\n",
    "\n",
    "mkdir $1/analysis/09_CIPHER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i DIR,patients\n",
    "\n",
    "colnames(patients) <- c(\"patient\", \"normal\", \"tumor\")\n",
    "patients_full <- patients %>% pivot_longer(cols = !patient, names_to = \"normal_tumor\", values_to = \"sample\")\n",
    "\n",
    "all <- data.frame(transcript_id = character(),\n",
    "                  transcript_type = factor(), \n",
    "                  transcript_name = character(),\n",
    "                  patient = factor(),\n",
    "                  stringsAsFactors=FALSE)\n",
    "\n",
    "for (i in 1:nrow(patients)) {\n",
    "  \n",
    "  df <- as.data.frame(rtracklayer::import(paste0(DIR, \"/analysis/08_tumor_specific/\",patients[i,1],\"/\",patients[i,1],\"_known_tumor_specific_genes_1FPKM_300kb_CDS.gtf\")))\n",
    "  df <- df[df$type == \"transcript\",]\n",
    "  df <- df[df$transcript_type == \"protein_coding\",] #only the protein coding ones at transcript level\n",
    "  df$patient <- patients[i,1]\n",
    "  df_selected <- df %>% select(transcript_id, gene_id, gene_name)\n",
    "  \n",
    "  all <- rbind(all,df_selected)\n",
    "}  \n",
    "\n",
    "times <- all %>% count(transcript_id)\n",
    "times_ordered <- times[order(times$n, decreasing=TRUE),]\n",
    "print(times_ordered %>% head)\n",
    "\n",
    "total <- merge(times_ordered,all, by=\"transcript_id\")\n",
    "total <- total[,c(1,3,4,2)]\n",
    "total <- total[order(total$n, decreasing=TRUE),]\n",
    "total_unique <- total %>% distinct()\n",
    "total_unique$transcript_id <- sub(\"*\\\\.[0-9]\", \"\", total_unique$transcript_id)\n",
    "total_unique$gene_id <- sub(\"*\\\\.[0-9]\", \"\", total_unique$gene_id)\n",
    "total_unique %>% head\n",
    "\n",
    "write.csv(total_unique, file.path(DIR,\"analysis/09_CIPHER/common_coding_genes.csv\"),row.names = FALSE, quote = FALSE)\n",
    "\n",
    "ggplot(times, aes(x=n)) +\n",
    "  geom_bar(fill=\"#636363\") +\n",
    "  geom_text(stat=\"count\", aes(label=..count..), vjust=-1, size = 2.5) +\n",
    "  ggtitle(\"Common non-coding genes across patients | lncRNA & processed_pseudogenes\") +\n",
    "  theme(axis.title.x = element_blank(), axis.title.y = element_blank(), plot.title = element_text(face=\"bold\", size = 9), legend.position = \"none\") \n",
    "ggsave(file.path(DIR,\"results/plots/common_coding_genes_patients.png\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GFFCOMPARE\n",
    "\n",
    "Create only one newly assembled tumor-specific transcriptome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#first modify the idenifiers so we can get information of the patient they belong only looking to the identifier\n",
    "try:\n",
    "    os.makedirs(DIR+\"/analysis/gffcompare\")\n",
    "except:\n",
    "    print(\"gffcompare directory already exists for dataset: %s\" %(PROJECT))\n",
    "    \n",
    "for p in patients_id:\n",
    "    file=DIR+\"/analysis/08_tumor_specific/\"+str(p)+\"/\"+str(p)+\"_novel_tumor_specific_genes_1FPKM_300kb.gtf\"\n",
    "    with open (file) as inp:\n",
    "        outfile=DIR+\"/analysis/gffcompare/\"+str(p)+\"_novel_tumor_specific_genes_1FPKM_300kb_MODIF.gtf\"\n",
    "        with open(outfile, 'w') as out:\n",
    "            lines=inp.readlines()\n",
    "            for line in lines:\n",
    "                final_line = line.replace('STRG', 'STRG.'+str(p))\n",
    "                out.write(final_line)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$DIR\" \"$patients_summary\"\n",
    "\n",
    "declare -a listfiles\n",
    "\n",
    "cat $2 | while read patient normal tumor; do\n",
    "    listfiles=(${listfiles[@]} $1/analysis/08_tumor_specific/$patient/${patient}_novel_tumor_specific_genes_1FPKM_300kb.gtf) # Add new element at the end of the array\n",
    "done \n",
    "\n",
    "cd $1/analysis/gffcompare\n",
    "\n",
    "printf \"%s\\n\" \"${listfiles[@]}\" > list_novel_tumorsp_files.txt\n",
    "\n",
    "module load gffcompare/20190227\n",
    "\n",
    "gffcompare -i list_novel_tumorsp_files.txt \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 12. Hepatocarcinoma Biomarkers\n",
    "\n",
    "**Gene expression downregulation**\n",
    "\n",
    "- MT1M/ENSG00000205364.4 (Mao et al., 2012, Carcinogenesis 33 : 2568–2577). \n",
    "\n",
    "- CDKN2A/ENSG00000147889.18 (The Cancer Genome Atlas Research Network, 2017, Cell 169, 1327–1341. Comprehensive and Integrative Genomic Characterization of Hepatocellular Carcinoma).\n",
    "\n",
    "\n",
    "**Gene expression upregulation**\n",
    "\n",
    "- TERT/ENSG00000164362.21 (The Cancer Genome Atlas Research Network, 2017, Cell 169, 1327–1341. Comprehensive and Integrative Genomic Characterization of Hepatocellular Carcinoma).\n",
    "\n",
    "- GABRD/ENSG00000187730.9, THBDS4/ENSG00000113296.14 (!THBS4) (Sun et al., 2018. An integrated analysis of genome-wide DNA methylation and gene expression data in hepatocellular carcinoma. FEBS Open Bio).\n",
    "\n",
    "- LINC01419/ENSG00000253898.1, HAGLR/ENSG00000224189.8, LINC00176/ENSG00000196421.8 (Unfried et al., 2019. Identification of Coding and Long Noncoding RNAs Differentially Expressed in Tumors and Preferentially Expressed in Healthy Tissues. Cancer Res).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "biomarkers=pd.read_csv(os.path.join(GENOMEDIR,\"hepatocarcinoma_biomarkers.txt\"))\n",
    "try:\n",
    "    os.mkdir(DIR+\"/analysis/12_biomarkers\")\n",
    "except:\n",
    "    print(\"12_biomarkers directory already exists, some data may be replaced\")\n",
    "\n",
    "for p in patients_id:\n",
    "    print(p)\n",
    "    present = pd.DataFrame()\n",
    "    merged_df = pd.DataFrame()\n",
    "    full_df = pd.DataFrame()\n",
    "    INDIR=DIR+\"/analysis/08_tumor_specific/\"+p+\"/\"\n",
    "    file=p+\"_known_tumor_specific_genes_01_1FPKM_300kb.csv\"\n",
    "    df=pd.read_csv(os.path.join(INDIR,file), names=['gene_id','FPKM','Length','known_novel']) \n",
    "   \n",
    "    out=DIR+\"/analysis/12_biomarkers/\"+p+\"_hepato_biomarkers.csv\" \n",
    "    for each in biomarkers['gene_id']:\n",
    "        if df['gene_id'].str.contains(each).any():\n",
    "            present = present.append(df[df['gene_id'] == each], ignore_index=True)\n",
    "            merged_df = pd.merge(present,biomarkers.loc[biomarkers['gene_id'] == each], on = \"gene_id\")\n",
    "            full_df=full_df.append(merged_df, ignore_index=False)\n",
    "    full_df.to_csv(out, index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i GENOMEDIR\n",
    "\n",
    "colnames(patients) <- c(\"patient\", \"normal\", \"tumor\")\n",
    "patients_full <- patients %>% pivot_longer(cols = !patient, names_to = \"normal_tumor\", values_to = \"sample\")\n",
    "\n",
    "biomarkers <- read.csv(paste0(GENOMEDIR,\"/hepatocarcinoma_biomarkers.txt\"))\n",
    "\n",
    "full <- data.frame(gene_id = character(),\n",
    "                      sample = factor(),\n",
    "                      Length = numeric(),\n",
    "                      FPKM = numeric(),\n",
    "                   normal_tumor = factor(),\n",
    "                      patient=factor(),\n",
    "                   gene_name = character(),\n",
    "                   type=factor(),\n",
    "                      stringsAsFactors = FALSE)\n",
    "\n",
    "for (i in 1:nrow(patients)) {\n",
    "  csv <- read.csv(paste0(DIR,\"/analysis/07_quantification/\",patients[i,1],\"/\",patients[i,1],\"_table_of_counts_300kb_RPKM.csv\"))\n",
    "  names(csv)[1] <- \"gene_id\"\n",
    "  csv_long <- csv %>% pivot_longer(cols = starts_with('SRR'), names_to = \"sample\", values_to = \"FPKM\")\n",
    "  csv_long <- merge(csv_long, patients_full, by=\"sample\")\n",
    "  biomarkers_in_sample <- csv_long[csv_long$gene_id %in% biomarkers$gene_id,]\n",
    "  biomarkers_in_sample <- merge(biomarkers_in_sample, biomarkers, by=\"gene_id\")\n",
    "  biomarkers_in_sample$logFPKM <- log10(biomarkers_in_sample$FPKM)\n",
    "\n",
    "  full <- rbind(full,biomarkers_in_sample)\n",
    "}\n",
    "\n",
    "write.csv(full, paste0(DIR,\"/analysis/12_biomarkers/total_biomarkers_values.csv\"), row.names=FALSE, quote=FALSE)\n",
    "ggplot(full, aes(x=gene_name, y=logFPKM)) +\n",
    "geom_boxplot(aes(fill=normal_tumor)) +\n",
    "geom_point(position=position_jitterdodge(jitter.width=1, dodge.width = 0), \n",
    "         pch=21, aes(fill=normal_tumor), show.legend = F, size=1) +\n",
    "scale_fill_manual(\"Adjacent or Tumor sample\", values=c(\"#31a354\", \"#de2d26\")) +\n",
    "ggtitle(\"Gene expression cancer biomarkers\") +\n",
    "  theme(axis.text.x = element_text(angle = 45, hjust=1), plot.title = element_text(face=\"bold\", size = 10))\n",
    "ggsave(paste0(DIR,\"/results/plots/cancer_markers_expression_all.png\"), height=12, width=8)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
